### DISCUSSION TITLE
How to Change Magic Key of Pyinstaller Using BootLoader ?

### DISCUSSION THREAD

**Discussion by PremKumarMishra** [pyinstaller/pyinstaller#5988] [Help] [answered]
I am a user of Pyinstaller Since 3 years.I like pyinstaller most than any other exe makers Because of its simplicity.
But Recently I have figured out that Pyinstaller Executable can be unpacked by pyinextractor and pyc files can be decrypted by uncompyle6.So I wish i can change the magic key _**b'MEI\014\013\012\013\016'**_ by making custom bootloader and changing some other stuffs.Then pyinextractor will not recognize it as Pyinstaller Archive. Please Reply What to Change.
Thanks in Advance.

**Comment by rokm** [comment, marked as answer]
While you could change the MAGIC string in both [bootloader](https://github.com/pyinstaller/pyinstaller/blob/878454cdad89cb6a07e681a1c6dc5dcd21a10da3/bootloader/src/pyi_archive.c#L351) and [python-based CArchive writer](https://github.com/pyinstaller/pyinstaller/blob/878454cdad89cb6a07e681a1c6dc5dcd21a10da3/PyInstaller/archive/writers.py#L306), I doubt that would offer you any extra "protection". I mean, if someone wanted to decompile your program, they could simply check the binary for the new magic value and adjust the `pyiextractor` (or the underyling `pyinstaller`) code to handle that as well...

(And by the way, PyInstaller comes with its own [`pyi-archive_viewer` tool](https://github.com/pyinstaller/pyinstaller/blob/develop/archive_viewer.py) that allows extraction of data from the executable-embedded archive(s)...)

**Reply by PremKumarMishra** [reply]
I did try to change the string in bootloader and in pyinstaller archive/writer.
But it threw me error **Can not open myfile.archive or myfile.pkg**.
And Can you Suggest Your Extra Protection ? I will really appreciate your help.
I am making my product for Clients So Code Protection is Important.

**Reply by PremKumarMishra** [reply]
![Capture](https://user-images.githubusercontent.com/74233905/124900609-df8a4480-dffe-11eb-9cee-4987dcabfb4a.PNG)
I tried to change the **MEI** Charecter to **SYS** in boootloader/pyiarchive.c
![Capture](https://user-images.githubusercontent.com/74233905/124900923-2841fd80-dfff-11eb-9f29-b4fb49c7fe91.PNG)
and changed **MEI** to **SYS** in MAGIC of pyinstaller/archive/writer.py
still giving error **Can not open myfile.archive or myfile.pkg.**
Thanks For Replying in Advance

**Reply by rokm** [reply]
Have you recompiled the bootloader, reinstalled PyInstaller, and rebuilt your program?

Changing the MAGIC in the indicated places works for me...

**Reply by rokm** [reply]
> And Can you Suggest Your Extra Protection ? I will really appreciate your help.

See #5687 for discussion regarding similar concerns.

> I am making my product for Clients So Code Protection is Important.

The way I see it, if your client is motivated enough to reverse-engineer your code, you will have to go a long way to prevent them from doing so. And simple obfuscation tricks like the one you're currently considering won't do you much good against that. PyInstaller's internal structures, while not documented, are not exactly classified, either.

**Reply by bwoodsend** [reply]
I think the fact that you yourself have come up with the idea of changing the MEI string without suggestion from one of the PyInstaller developers indicates that the MEI string is still reasonably well known.

**Reply by PremKumarMishra** [reply]
> Have you recompiled the bootloader, reinstalled PyInstaller, and rebuilt your program?
> 
> Changing the MAGIC in the indicated places works for me...

I only Compiled bootloader to exe and pasted in Bootloader/Windows32 and Windows64 folder 
and saved the writers.py script

**Reply by PremKumarMishra** [reply]
> I think the fact that you yourself have come up with the idea of changing the MEI string without suggestion from one of the PyInstaller developers indicates that the MEI string is still reasonably well known.

Then Can you suggest me by changing which thing it will help me more in protection case.

**Reply by PremKumarMishra** [reply]
> > And Can you Suggest Your Extra Protection ? I will really appreciate your help.
> 
> See #5687 for discussion regarding similar concerns.
> 
> > I am making my product for Clients So Code Protection is Important.
> 
> The way I see it, if your client is motivated enough to reverse-engineer your code, you will have to go a long way to prevent them from doing so. And simple obfuscation tricks like the one you're currently considering won't do you much good against that. PyInstaller's internal structures, while not documented, are not exactly classified, either.

i saw the cython pyd .I am thinking to try but still i dont want exe to be extracted first of all.
If thats possible then it would be great.

**Reply by bwoodsend** [reply]
Would nothing mentioned in #5687 do?

**Reply by bwoodsend** [reply]
Personally I think I'd go for [this one](http://liftoff.github.io/pyminifier/#special-sauce) just because I find the idea of someone trying to read random unicode variables funny.

**Reply by bwoodsend** [reply]
All of them are going to make diagnostics pretty hellish though...

**Reply by rokm** [reply]
> > Have you recompiled the bootloader, reinstalled PyInstaller, and rebuilt your program?
> > Changing the MAGIC in the indicated places works for me...
> 
> I only Compiled bootloader to exe and pasted in Bootloader/Windows32 and Windows64 folder
> and saved the writers.py script

If it doesn't work, then you have a mismatch between bootloader and python code. It should be pretty easy to tell which one is mismatched, though, based on patterns in the .exe...

**Reply by PremKumarMishra** [reply]
> Would nothing mentioned in #5687 do?

That Does.That tells protecting soruce code through cython pyd

**Reply by bwoodsend** [reply]
Well let us know how it goes, whatever you do. We get asked this quite often so it'd be good to have a wiki page we can just point people to.

**Reply by PremKumarMishra** [reply]
> Well let us know how it goes, whatever you do. We get asked this quite often so it'd be good to have a wiki page we can just point people to.

i did a look at pyminifier but only you can run encrypted py file by making pyz . Now how i will compile pyz file through Pyinstaller

**Reply by PremKumarMishra** [reply]
> > > Have you recompiled the bootloader, reinstalled PyInstaller, and rebuilt your program?
> > > Changing the MAGIC in the indicated places works for me...
> > 
> > 
> > I only Compiled bootloader to exe and pasted in Bootloader/Windows32 and Windows64 folder
> > and saved the writers.py script
> 
> If it doesn't work, then you have a mismatch between bootloader and python code. It should be pretty easy to tell which one is mismatched, though, based on patterns in the .exe...

so how can i check what is mismatched. Any idea ?

**Reply by bwoodsend** [reply]
> i did a look at pyminifier but only you can run encrypted py file by making pyz . Now how i will compile pyz file through Pyinstaller

That's only if you use its zlib compression feature. That won't provide any protection. I've just used:
```
pyminifier --nonlatin --replacement-length=10 -O -o mangled.py --obfuscate-builtins --obfuscate-import-methods tumult.py
```
to convert the example from pyminifier's page (called `tumult.py`) to `mangled.py` with all the obfuscation options enabled. You can just invoke PyInstaller on `mangled.py` which looks like:
```python
#!/usr/bin/env python
𬒹𠴅俻𣝞𤎼𐠴𐪄𐬍𐪂㷦=ImportError
𬒹𠴅俻𣝞𤎼𐠴𐪄𐬍𐪂놛=print
𬒹𠴅俻𣝞𤎼𐠴𐪄𐬍𐪂ލ=False
𬒹𠴅俻𣝞𤎼𐠴𐪄𐬍𐪂ﹻ=object
try:
 import demiurgic
except 𬒹𠴅俻𣝞𤎼𐠴𐪄𐬍𐪂㷦:
 𬒹𠴅俻𣝞𤎼𐠴𐪄𐬍𐪂놛("Warning: You're not demiurgic. Actually, I think that's normal.")
try:
 import mystificate
except 𬒹𠴅俻𣝞𤎼𐠴𐪄𐬍𐪂㷦:
 𬒹𠴅俻𣝞𤎼𐠴𐪄𐬍𐪂놛("Warning: Dark voodoo may be unreliable.")
𬒹𠴅俻𣝞𤎼𐠴𐪄𐬍𐪂ﺄ=𬒹𠴅俻𣝞𤎼𐠴𐪄𐬍𐪂ލ 
class 𬒹𠴅俻𣝞𤎼𐠴𐪄𐬍𐪂ݰ(𬒹𠴅俻𣝞𤎼𐠴𐪄𐬍𐪂ﹻ):
 def __init__(self,*args,**kwargs):
  pass
 def 𬒹𠴅俻𣝞𤎼𐠴𐪄𐬍𐪂䤄(self,dactyl):
  𬒹𠴅俻𣝞𤎼𐠴𐪄𐬍𐪂ﮍ=demiurgic.palpitation(dactyl)
  𬒹𠴅俻𣝞𤎼𐠴𐪄𐬍𐪂ﲴ=mystificate.dark_voodoo(𬒹𠴅俻𣝞𤎼𐠴𐪄𐬍𐪂ﮍ)
  return 𬒹𠴅俻𣝞𤎼𐠴𐪄𐬍𐪂ﲴ
 def 𬒹𠴅俻𣝞𤎼𐠴𐪄𐬍𐪂䒂(self,whatever):
  𬒹𠴅俻𣝞𤎼𐠴𐪄𐬍𐪂놛(whatever)
if __name__=="__main__":
 𬒹𠴅俻𣝞𤎼𐠴𐪄𐬍𐪂놛("Forming...")
 𬒹𠴅俻𣝞𤎼𐠴𐪄𐬍𐪂ਝ=𬒹𠴅俻𣝞𤎼𐠴𐪄𐬍𐪂ݰ("epicaricacy","perseverate")
 𬒹𠴅俻𣝞𤎼𐠴𐪄𐬍𐪂ਝ.𬒹𠴅俻𣝞𤎼𐠴𐪄𐬍𐪂䒂("Codswallop")
# Created by pyminifier (https://github.com/liftoff/pyminifier)
```

Chances are you'll want to make a copy of all your code, then mangle every file in-place with the same name or your imports will get mixed up but I'll leave that to you.

**Reply by rokm** [reply]
> > > > Have you recompiled the bootloader, reinstalled PyInstaller, and rebuilt your program?
> > > > Changing the MAGIC in the indicated places works for me...
> > > 
> > > 
> > > I only Compiled bootloader to exe and pasted in Bootloader/Windows32 and Windows64 folder
> > > and saved the writers.py script
> > 
> > 
> > If it doesn't work, then you have a mismatch between bootloader and python code. It should be pretty easy to tell which one is mismatched, though, based on patterns in the .exe...
> 
> so how can i check what is mismatched. Any idea ?

Check the strings in the exe, which should contain two copies of your new MAGIC string. The first one should correspond to bootloader's code (i.e., the search pattern encoded in bootloader), and the second one (located towards (but not exactly at) the end of the file) should correspond to the one in CArchive that python code generated.

**Reply by PremKumarMishra** [reply]
> > > > > Have you recompiled the bootloader, reinstalled PyInstaller, and rebuilt your program?
> > > > > Changing the MAGIC in the indicated places works for me...
> > > > 
> > > > 
> > > > I only Compiled bootloader to exe and pasted in Bootloader/Windows32 and Windows64 folder
> > > > and saved the writers.py script
> > > 
> > > 
> > > If it doesn't work, then you have a mismatch between bootloader and python code. It should be pretty easy to tell which one is mismatched, though, based on patterns in the .exe...
> > 
> > 
> > so how can i check what is mismatched. Any idea ?
> 
> Check the strings in the exe, which should contain two copies of your new MAGIC string. The first one should correspond to bootloader's code (i.e., the search pattern encoded in bootloader), and the second one (located towards (but not exactly at) the end of the file) should correspond to the one in CArchive that python code generated.

Thanks I will must check it.

**Reply by PremKumarMishra** [reply]
> > i did a look at pyminifier but only you can run encrypted py file by making pyz . Now how i will compile pyz file through Pyinstaller
> 
> That's only if you use its zlib compression feature. That won't provide any protection. I've just used:
> 
> ```
> pyminifier --nonlatin --replacement-length=10 -O -o mangled.py --obfuscate-builtins --obfuscate-import-methods tumult.py
> ```
> 
> to convert the example from pyminifier's page (called `tumult.py`) to `mangled.py` with all the obfuscation options enabled. You can just invoke PyInstaller on `mangled.py` which looks like:
> 
> ```python
> #!/usr/bin/env python
> 𬒹𠴅俻𣝞𤎼𐠴𐪄𐬍𐪂㷦=ImportError
> 𬒹𠴅俻𣝞𤎼𐠴𐪄𐬍𐪂놛=print
> 𬒹𠴅俻𣝞𤎼𐠴𐪄𐬍𐪂ލ=False
> 𬒹𠴅俻𣝞𤎼𐠴𐪄𐬍𐪂ﹻ=object
> try:
>  import demiurgic
> except 𬒹𠴅俻𣝞𤎼𐠴𐪄𐬍𐪂㷦:
>  𬒹𠴅俻𣝞𤎼𐠴𐪄𐬍𐪂놛("Warning: You're not demiurgic. Actually, I think that's normal.")
> try:
>  import mystificate
> except 𬒹𠴅俻𣝞𤎼𐠴𐪄𐬍𐪂㷦:
>  𬒹𠴅俻𣝞𤎼𐠴𐪄𐬍𐪂놛("Warning: Dark voodoo may be unreliable.")
> 𬒹𠴅俻𣝞𤎼𐠴𐪄𐬍𐪂ﺄ=𬒹𠴅俻𣝞𤎼𐠴𐪄𐬍𐪂ލ 
> class 𬒹𠴅俻𣝞𤎼𐠴𐪄𐬍𐪂ݰ(𬒹𠴅俻𣝞𤎼𐠴𐪄𐬍𐪂ﹻ):
>  def __init__(self,*args,**kwargs):
>   pass
>  def 𬒹𠴅俻𣝞𤎼𐠴𐪄𐬍𐪂䤄(self,dactyl):
>   𬒹𠴅俻𣝞𤎼𐠴𐪄𐬍𐪂ﮍ=demiurgic.palpitation(dactyl)
>   𬒹𠴅俻𣝞𤎼𐠴𐪄𐬍𐪂ﲴ=mystificate.dark_voodoo(𬒹𠴅俻𣝞𤎼𐠴𐪄𐬍𐪂ﮍ)
>   return 𬒹𠴅俻𣝞𤎼𐠴𐪄𐬍𐪂ﲴ
>  def 𬒹𠴅俻𣝞𤎼𐠴𐪄𐬍𐪂䒂(self,whatever):
>   𬒹𠴅俻𣝞𤎼𐠴𐪄𐬍𐪂놛(whatever)
> if __name__=="__main__":
>  𬒹𠴅俻𣝞𤎼𐠴𐪄𐬍𐪂놛("Forming...")
>  𬒹𠴅俻𣝞𤎼𐠴𐪄𐬍𐪂ਝ=𬒹𠴅俻𣝞𤎼𐠴𐪄𐬍𐪂ݰ("epicaricacy","perseverate")
>  𬒹𠴅俻𣝞𤎼𐠴𐪄𐬍𐪂ਝ.𬒹𠴅俻𣝞𤎼𐠴𐪄𐬍𐪂䒂("Codswallop")
> # Created by pyminifier (https://github.com/liftoff/pyminifier)
> ```
> 
> Chances are you'll want to make a copy of all your code, then mangle every file in-place with the same name or your imports will get mixed up but I'll leave that to you.

Thanks For This Trick.

**Reply by PremKumarMishra** [reply]
> Have you recompiled the bootloader, reinstalled PyInstaller, and rebuilt your program?
> 
> Changing the MAGIC in the indicated places works for me...

I know i marked question as answer but still i could not fix that .But here you said it works for you.
Can you please give me the bootloader files and writer.py you made for last time ?
Thanks For All Your Help.

**Reply by rokm** [reply]
>  Can you please give me the bootloader files and writer.py you made for last time ?

Here you go: https://github.com/rokm/pyinstaller/tree/modified-magic-example

**Reply by PremKumarMishra** [reply]
> > Can you please give me the bootloader files and writer.py you made for last time ?
> 
> Here you go: https://github.com/rokm/pyinstaller/tree/modified-magic-example

Thanks A Lot
